### PR TITLE
Babe integration

### PR DESCRIPTION
**Pull request by drgora**
This PR updates the consensus to drop Aura block authoring in favor of Babe.

Runtime tests are updated and integrated to account for the new pallet, and its interactions.

### CODE REVIEW

File: runtime/src/tests.rs [hunk 1]
```diff
@@ -173,37 +189,72 @@ fn pallet_poe_availability() {
...
+    // Any random values for these constants should do
+    const BLOCK_NUMBER: BlockNumber = 1;
+    const SLOT_ID: u64 = 87;
+    const BABE_AUTHOR_ID: u32 = 1;
+
+    // Initialize block #BLOCK_NUMBER, authored at slot SLOT_ID by BABE_AUTHOR_ID using Babe
+    fn initialize() {
+        let slot = Slot::from(SLOT_ID);
+        let authority_index = BABE_AUTHOR_ID;
+        let transcript = sp_consensus_babe::VrfTranscript::new(b"test", &[]);
```
Comment (la10736) [inline runtime/src/tests.rs:201]: You're using this function to init the chain in almost all following tests. Consider to redefine `new_test_ex()` function to include this code too:

```rust
    fn new_test_ext() -> sp_io::TestExternalities {
        let mut ex = super::new_test_ext();
        ex.execute_with(|| initialize());
        ex
    }
```

File: node/src/chain_spec.rs [hunk 2]
```diff
@@ -148,6 +148,7 @@ pub fn local_config() -> Result<ChainSpec, String> {
     )
     .with_name("NH Local")
     .with_id("nh_local")
+    .with_protocol_id("nh_local")
```
Comment (la10736) [inline node/src/chain_spec.rs:151]: We can use `zen` or `tzen` or `nzen` .... seams more close to what polkadot do

File: node/src/chain_spec.rs [hunk 3]
```diff
@@ -181,6 +182,7 @@ pub fn testnet_config() -> Result<ChainSpec, String> {
     )
     .with_name("NH Testnet")
     .with_id("nh_testnet")
+    .with_protocol_id("nh_testnet")
```
Comment (la10736) [inline node/src/chain_spec.rs:185]: Here we can use `tzen`

**Review by la10736 — APPROVED** [review]
LGTM

File: runtime/src/tests.rs (hunk 1 above)
Comment (drgora) [reply runtime/src/tests.rs:201]: Good idea, done!

File: node/src/chain_spec.rs (hunk 2 above)
Comment (drgora) [reply node/src/chain_spec.rs]: Done

File: node/src/chain_spec.rs (hunk 3 above)
Comment (drgora) [reply node/src/chain_spec.rs]: Done

**Comment by JackPiri** [comment]
Please modify entrypoint.sh as suggested below:
```
  echo "Injecting key (Babe)"
  ${NH_NODE} key insert "${injection_args[@]}" \
    --scheme Sr25519 \
    --suri "${NH_SECRET_PHRASE_PATH}" \
    --key-type babe
```
(reference: https://github.com/paritytech/polkadot-sdk/blob/master/substrate/primitives/core/src/crypto.rs)

File: runtime/src/tests.rs [hunk 4]
```diff
@@ -18,27 +18,35 @@ use super::*;
...
     TPublic::Pair::from_string(&format!("//test_seed{}", seed), None)
         .expect("static values are valid; qed")
-        .public()
 }
 
-pub const SLOT_ID: u64 = 87; // Any random value should do
-pub const NUM_VALIDATORS: u32 = 2;
-pub const AURA_AUTHOR_ID: u32 = (SLOT_ID as u32) % NUM_VALIDATORS;
+const NUM_VALIDATORS: u32 = 2;
+const TEST_PRIMARY_PROBABILITY: (u64, u64) = (1, 4);
```
[the remark this answers is not in the document]
Comment (drgora) [reply runtime/src/tests.rs]: Yes, makes sense, done

File: runtime/src/lib.rs [hunk 5]
```diff
@@ -352,7 +363,7 @@ impl pallet_staking::EraPayout<Balance> for ZendPayout {
 }
 
 parameter_types! {
-    pub const SessionsPerEra: sp_staking::SessionIndex = 6 * HOURS / Period::get(); // number of sessions in 1 era, 6h
+    pub const SessionsPerEra: sp_staking::SessionIndex = 6 * HOURS / EpochDurationInBlocks::get(); // number of sessions in 1 era, 6h
```
[the remark this answers is not in the document]
Comment (drgora) [reply runtime/src/lib.rs:366]: That would be fine and equivalent in terms of era duration, but longer sessions might incur longer response times in session handling (for example if active validators misbehave and a new set should be activated).

**Comment by drgora** [comment]
> Please modify `entrypoint.sh` as suggested below:
> 
> ```
>   echo "Injecting key (Babe)"
>   ${NH_NODE} key insert "${injection_args[@]}" \
>     --scheme Sr25519 \
>     --suri "${NH_SECRET_PHRASE_PATH}" \
>     --key-type babe
> ```
> 
> (reference: https://github.com/paritytech/polkadot-sdk/blob/master/substrate/primitives/core/src/crypto.rs)

Done as suggested, thank you.

**Review by ptagl — APPROVED** [review]

File: runtime/src/lib.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (JackPiri) [reply runtime/src/lib.rs:366]: Ok we are fine in increasing the session length from 100 blocks to 600 blocks.

**Pull request merged**